pex/resolve/resolved_requirement.py

```diff
@@ -77,29 +78,72 @@ def from_hashing_fingerprint(cls, fingerprint):
 
 @attr.s(frozen=True)
 class ArtifactURL(object):
+    # These ranks prefer the highest digest size and then use alphabetic order for a tie-break.
+
+    _RANKED_ALGORITHMS = tuple(
+        sorted(
+            hashlib.algorithms_guaranteed,
+            key=lambda alg: (-hashlib.new(alg).digest_size, alg),
+        )
+    )
+
     @classmethod
     def parse(cls, url):
         # type: (str) -> ArtifactURL
         url_info = urlparse.urlparse(url)
+        scheme = parse_scheme(url_info.scheme) if url_info.scheme else None
+        path = url_unquote(url_info.path)
+
+        fingerprints = []
+        fragment_parameters = urlparse.parse_qs(url_info.fragment)
+        if fragment_parameters:
+            # Artifact URLs from indexes may contain pre-computed hashes. We isolate those here,
+            # centrally, if present.
+            # See: https://peps.python.org/pep-0503/#specification
+            for alg in cls._RANKED_ALGORITHMS:
+                hashes = fragment_parameters.pop(alg, None)
+                if not hashes:
+                    continue
+                if len(hashes) > 1 and len(set(hashes)) > 1:
+                    TRACER.log(
+                        "The artifact url contains multiple distinct hash values for the {alg} "
+                        "algorithm, not trusting any of these: {url}".format(alg=alg, url=url)
+                    )
+                    continue
+                fingerprints.append(Fingerprint(algorithm=alg, hash=hashes[0]))
+
+        download_url = urlparse.urlunparse(url_info._replace(params="", query="", fragment=""))
         normalized_url = urlparse.urlunparse(
-            (url_info.scheme, url_info.netloc, url_unquote(url_info.path).rstrip(), "", "", "")
+            url_info._replace(path=path, params="", query="", fragment="")
         )
         return cls(
             raw_url=url,
+            download_url=download_url,
             normalized_url=normalized_url,
-            scheme=parse_scheme(url_info.scheme) if url_info.scheme else None,
-            path=url_unquote(url_info.path),
+            scheme=scheme,
+            path=path,
+            fragment_parameters=fragment_parameters,
+            fingerprints=tuple(fingerprints),
         )
 
     raw_url = attr.ib(eq=False)  # type: str
+    download_url = attr.ib(eq=False)  # type: str
     normalized_url = attr.ib()  # type: str
-    scheme = attr.ib()  # type: Optional[Union[str, ArchiveScheme.Value, VCSScheme]]
+    scheme = attr.ib(eq=False)  # type: Optional[Union[str, ArchiveScheme.Value, VCSScheme]]
```

Comment:
Why is this necessary?

Reply:
It wasn't, but neither was its redundant role in equality as part of the normalized URL either.